ValveResourceFormat/Resource/ResourceTypes/KV3.cs

```diff
@@ -0,0 +1,592 @@
+        }
+
+        //header state
+        private static void ReadHeader(char c)
+        {
+            currentString.Append(c);
+
+            //Read until --> is encountered
+            if (c == '>' && currentString.ToString().Substring( currentString.Length - 3 ) == "-->") {
+                stateStack.Pop();
+                stateStack.Push(State.SEEK_VALUE);
+                return;
+            }
+        }
+
+        //Seeking value state
+        private static void SeekValue(char c)
+        {
+            //Ignore whitespace
+            if (whiteSpace.Contains(c) || c == '=' )
+            {
+                return;
+            }
+
+            //Check struct opening
+            if (c == '{')
+            {
+                stateStack.Pop();
+                stateStack.Push(State.VALUE_STRUCT);
+
+                objStack.Push(new KVObject(currentString.ToString()));
+            }
+            //Check for array opening
+            else if (c == '[')
+            {
+                stateStack.Pop();
+                stateStack.Push(State.VALUE_ARRAY);
+                stateStack.Push(State.SEEK_VALUE);
+
+                objStack.Push(new KVArray(currentString.ToString()));
+            }
+            //Check for array closing
+            else if (c == ']')
+            {
+                stateStack.Pop();
+                stateStack.Pop();
+
+                KVObject value = objStack.Pop();
+                objStack.Peek().AddProperty(value.key, new KVValue(typeof(KVObject),value));
+            }
+            //Multiline string opening
+            else if (c == '"' && fileStream[index + 1] == '"' && fileStream[index + 2] == '"')
+            {
+                stateStack.Pop();
+                stateStack.Push(State.VALUE_STRING_MULTI);
+                currentString.Clear();
+
+                //Skip to the start of the string
+                index += 2;
+            }
+            //String opening
+            else if (c == '"')
+            {
+                stateStack.Pop();
+                stateStack.Push(State.VALUE_STRING);
+                currentString.Clear();
+            }
+            //Booleans
+            else if ((c == 'f' && (fileStream[index + 1] == 'a' && fileStream[index + 2] == 'l' && fileStream[index + 3] == 's' && fileStream[index + 4] == 'e'))
+                || (c == 't' && (fileStream[index + 1] == 'r' && fileStream[index + 2] == 'u' && fileStream[index + 3] == 'e')))
+            {
+                stateStack.Pop();
+                stateStack.Push(State.VALUE_BOOL);
+                currentString.Clear();
+                currentString.Append(c);
+            }
+            //Number
+            else if (numerals.Contains(c))
+            {
+                stateStack.Pop();
+                stateStack.Push(State.VALUE_NUMBER);
+                currentString.Clear();
+                currentString.Append(c);
+            }
+            //Flagged resource
```

Comment:
Code doesn't explain what a flagged resource is, just that it's a `FlaggedString` or a `string` - I'm unclear on what this is, judging by this PR, and having little experience with KeyValues3. Either explain with a comment, or let the code do the talking.

Also, is it guaranteed for all eternity that anything that is not the above is a flagged resource, or is there some identifier you can use to detect a flagged resource specifically?